### PR TITLE
Add IStartUp back /4

### LINKED ISSUE #27844: Custom WebApplicationFactory doesn't work correctly for overriding Configure Method in ConfigureWebHost
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Describe the bug

Hi,
I've created a custom [WebApplicationFactory](https://github.com/mehdihadeli/WebApplicationFactoryWithHost-Sample/blob/master/src/Test/WebApplicationFactoryWithHost.cs) and in this factory I override [ConfigureWebHost](https://github.com/mehdihadeli/WebApplicationFactoryWithHost-Sample/blob/bug/src/Test/WebApplicationFactoryWithHost.cs#L31) and inner this `WebHostBuilder` I have a [Configure](https://github.com/mehdihadeli/WebApplicationFactoryWithHost-Sample/blob/bug/src/Test/WebApplicationFactoryWithHost.cs#L45) for extending startup configuration. But my problem is my [Startup Configure method](https://github.com/mehdihadeli/WebApplicationFactoryWithHost-Sample/blob/bug/src/Test/SampleMiddlewareTests.cs#L40) doesn't call at all, and I get error because my startup Configure didn't call at first and when I remove [override Configure](https://github.com/mehdihadeli/WebApplicationFactoryWithHost-Sample/blob/bug/src/Test/WebApplicationFactoryWithHost.cs#L45) method in CustomFactory it works. I don't know Why?

Also, I defined a custom [CreateHostBuilder](https://github.com/mehdihadeli/WebApplicationFactoryWithHost-Sample/blob/bug/src/Test/WebApplicationFactoryWithHost.cs#L52) here.

### Expected Behavior

I want my [override Configure](https://github.com/mehdihadeli/WebApplicationFactoryWithHost-Sample/blob/master/src/Test/WebApplicationFactoryWithHost.cs#L30) method run after [startup Configure method](https://github.com/mehdihadeli/WebApplicationFactoryWithHost-Sample/blob/master/src/Test/SampleMiddlewareTests.cs#L41)

### Steps To Reproduce

For producing error, just run [this test](https://github.com/mehdihadeli/WebApplicationFactoryWithHost-Sample/blob/bug/src/Test/SampleMiddlewareTests.cs#L30).

### Exceptions (if any)

_No response_

### .NET Version

7.0.0

### Anything else?

*** EDIT by @Rick-Anderson  to add meta-data**

---
#### Document Details

⚠ *Do not edit this section. It is required for learn.microsoft.com ➟ GitHub issue linking.*

* ID: 793395fe-c738-22ce-d0e9-29a850c0fa88
* Version Independent ID: 0546ce3a-3258-9817-0f3f-13281c2f55da
* Content: [App startup in ASP.NET Core](https://learn.microsoft.com/en-us/aspnet/core/fundamentals/startup?view=aspnetcore-6.0&branch=pr-en-us-23615)
* Content Source: [aspnetcore/fundamentals/startup.md](https://github.com/dotnet/AspNetCore.Docs/blob/main/aspnetcore/fundamentals/startup.md)
* Product: **aspnet-core**
* Technology: **aspnetcore-fundamentals**
* GitHub Login: @Rick-Anderson
* Microsoft Alias: **riande**

---
[Associated WorkItem - 50838](https://dev.azure.com/msft-skilling/Content/_workitems/edit/50838)

### PR DESCRIPTION
**Pull request by Rick-Anderson**
Fixes #27844

[Internal review URL](https://review.learn.microsoft.com/en-us/aspnet/core/fundamentals/startup?view=aspnetcore-8.0&branch=pr-en-us-28090#extend-startup-with-startup-filters)

## Public review  of the doc build:
- download the [HTML files](https://www.dropbox.com/sh/yy994551cje0kap/AAB0tRnm7gze2kYrSYPSFHIda?dl=0)
- unzip
- view HTML file in a browser.
- For links outside the doc, remove the `review.` from URL.

### CODE REVIEW

File: aspnetcore/fundamentals/startup/7/WebStartup/Program.cs [hunk 1]
```diff
@@ -0,0 +1,26 @@
+using WebStartup.Middleware;
+
+var builder = WebApplication.CreateBuilder(args);
+
+builder.Services.AddRazorPages();
+builder.Services.AddTransient<IStartupFilter,
+                      RequestSetOptionsStartupFilter>();
```
Comment (Rick-Anderson) [inline aspnetcore/fundamentals/startup/7/WebStartup/Program.cs:7]: @mehdihadeli please review.

File: aspnetcore/test/integration-tests.md [hunk 2]
```diff
@@ -143,6 +143,8 @@ The `SendAsync` helper extension methods (`Helpers/HttpClientExtensions.cs`) and
 
 The [EF-Core in-memory database provider](/ef/core/testing/choosing-a-testing-strategy#in-memory-as-a-database-fake) can be used for limited and basic testing, however the ***[SQLite provider](/ef/core/testing/choosing-a-testing-strategy#sqlite-as-a-database-fake) is the recommended choice for in-memory testing***.
 
+See [Extend Startup with startup filters](xref:fundamentals/startup#IStartupFilter) which shows how to configure middleware using <xref:Microsoft.AspNetCore.Hosting.IStartupFilter>, which is useful when a test requires a custom service or middleware.
```
Comment (Rick-Anderson) [inline aspnetcore/test/integration-tests.md:146]: @mehdihadeli please review and suggest better wording. Is there a better place to add this? I added this to the end of [this section](https://learn.microsoft.com/en-us/aspnet/core/test/integration-tests?view=aspnetcore-7.0#customize-webapplicationfactory)

**Review by mehdihadeli — APPROVED** [review]
Seams great. Thanks a lot :)

File: aspnetcore/test/integration-tests.md (hunk 2 above)
Comment (mehdihadeli) [reply aspnetcore/test/integration-tests.md:146]: Yes, it is fine. Thanks

**Review by mehdihadeli — COMMENTED** [review]
Could we mention also #27844?

**Comment by Rick-Anderson** [comment]
@halter73 please recommend a reviewer.

File: aspnetcore/fundamentals/startup/includes/startup56.md [hunk 3]
```diff
@@ -0,0 +1,188 @@
...
+
+* Multiple `IStartupFilter` implementations may interact with the same objects. If ordering is important, order their `IStartupFilter` service registrations to match the order that their middlewares should run.
+* Libraries may add middleware with one or more `IStartupFilter` implementations that run before or after other app middleware registered with `IStartupFilter`. To invoke an `IStartupFilter` middleware before a middleware added by a library's `IStartupFilter`:
+
+  * Position the service registration before the library is added to the service container.
+  * To invoke afterward, position the service registration after the library is added.
+
+## Add configuration at startup from an external assembly
+
+An <xref:Microsoft.AspNetCore.Hosting.IHostingStartup> implementation allows adding enhancements to an app at startup from an external assembly outside of the app's `Startup` class. For more information, see <xref:fundamentals/configuration/platform-specific-configuration>.
```
Comment (halter73) [inline aspnetcore/fundamentals/startup/includes/startup56.md:178]: IHostingStartup also works with WebApplicationBuilder.

File: aspnetcore/fundamentals/startup.md [hunk 4]
```diff
@@ -170,16 +66,8 @@ Middleware execution order is set by the order of `IStartupFilter` registrations
   * Position the service registration before the library is added to the service container.
   * To invoke afterward, position the service registration after the library is added.
 
-## Add configuration at startup from an external assembly
-
-An <xref:Microsoft.AspNetCore.Hosting.IHostingStartup> implementation allows adding enhancements to an app at startup from an external assembly outside of the app's `Startup` class. For more information, see <xref:fundamentals/configuration/platform-specific-configuration>.
```
Comment (halter73) [inline aspnetcore/fundamentals/startup.md:175]: I know we cannot access the app's `Startup`, but can we still mention IHostingStartup an link to https://learn.microsoft.com/en-us/aspnet/core/fundamentals/host/platform-specific-configuration?view=aspnetcore-7.0 in the .NET 7 docs? That should still work with WebApplicationBuilder

File: aspnetcore/fundamentals/startup/includes/startup56.md [hunk 5]
```diff
@@ -0,0 +1,188 @@
...
+
+* Multiple `IStartupFilter` implementations may interact with the same objects. If ordering is important, order their `IStartupFilter` service registrations to match the order that their middlewares should run.
+* Libraries may add middleware with one or more `IStartupFilter` implementations that run before or after other app middleware registered with `IStartupFilter`. To invoke an `IStartupFilter` middleware before a middleware added by a library's `IStartupFilter`:
+
+  * Position the service registration before the library is added to the service container.
+  * To invoke afterward, position the service registration after the library is added.
+
+## Add configuration at startup from an external assembly
+
+An <xref:Microsoft.AspNetCore.Hosting.IHostingStartup> implementation allows adding enhancements to an app at startup from an external assembly outside of the app's `Startup` class. For more information, see <xref:fundamentals/configuration/platform-specific-configuration>. `IHostingStartup` works with <xref:Microsoft.AspNetCore.Builder.WebApplicationBuilder>.
```
Comment (halter73) [inline aspnetcore/fundamentals/startup/includes/startup56.md:178]: It doesn't make sense to mention `WebApplicationBuilder` in the .NET 5 docs. I also don't see why the .NET 6 and .NET 7 startup docs should be any different from each other.

**Review by halter73 — APPROVED** [review]

**Comment by halter73** [comment]
Sorry for the delay.

**Pull request merged**